packages/cli-kit/src/public/node/api/graphql.ts

```diff
@@ -114,13 +146,52 @@ async function performGraphQLRequest<TResult>(options: PerformGraphQLRequestOpti
     }
   }
 
+  const simpleUnauthorizedHandler = unauthorizedHandler?.type === 'simple' ? unauthorizedHandler.handler : undefined
+  const tokenRefreshHandler = unauthorizedHandler?.type === 'token_refresh' ? unauthorizedHandler.handler : undefined
+
+  const tokenRefreshUnauthorizedHandlerFunction = tokenRefreshHandler
+    ? async () => {
+        const refreshTokenResult = await tokenRefreshHandler()
+        if (refreshTokenResult.token) {
+          const {client: newClient, headers: newHeaders} = await createGraphQLClient({
+            url,
+            addedHeaders,
+            token: refreshTokenResult.token,
+          })
+          client = newClient
+          headers = newHeaders
+          return true
+        } else {
+          return false
+        }
+      }
+    : undefined
+
   const executeWithTimer = () =>
     runWithTimer('cmd_all_timing_network_ms')(async () => {
-      const response = await retryAwareRequest(
-        {request: performRequest, url, ...requestBehaviour},
-        responseOptions?.handleErrors === false ? undefined : errorHandler(api),
-        unauthorizedHandler,
-      )
+      let response
+      try {
+        response = await retryAwareRequest(
+          {request: performRequest, url, ...requestBehaviour},
+          responseOptions?.handleErrors === false ? undefined : errorHandler(api),
+          simpleUnauthorizedHandler,
+        )
+      } catch (error) {
+        if (error instanceof ClientError && error.response.status === 401 && tokenRefreshUnauthorizedHandlerFunction) {
+          const shouldRetry = await tokenRefreshUnauthorizedHandlerFunction()
+          if (shouldRetry) {
+            response = await retryAwareRequest(
```

Comment:
Can we combine this with the above, maybe a loop instead? It looks like they're separate because something different is happening, but they're intended to be the same.